src/main/java/org/candlepin/pki/impl/JSSProviderLoader.java

```diff
@@ -134,10 +161,73 @@ else if (e.getCause() instanceof KeyDatabaseException ||
             }
         }
 
-        jssProvider = new JSSProvider();
-        int addProviderReturn = Security.addProvider(jssProvider);
-        log.debug("Finished call to JSSProviderLoader.addProvider(). Returned value: {}",
-            addProviderReturn);
+        // Create a JSS provider to return
+        provider = new JSSProvider();
+
+        // Ensure the provider is not installed on the provider chain
+        if (Security.getProvider(PROVIDER_NAME) != null) {
+            log.warn("JSS security provider installed on provider chain; removing...");
+
+            // Don't pollute the provider space with JSS -- it's broken
+            Security.removeProvider(PROVIDER_NAME);
+        }
+
+        log.info("JSS initialization complete");
+    }
+
+    /**
+     * Fetches the JSS security provider. If JSS has not yet been initialized, and the initialize
+     * argument is true, it will be initialized before fetching the provider. If the initialize
+     * argument is false and JSS is not initialized, this method throws an exception.
+     *
+     * @param initialize
+     *  whether or not to initialize JSS if it has not yet been initialized
+     *
+     * @return
+     *  the JSS security provider
+     */
+    public static synchronized Provider getProvider(boolean initialize) {
+        if (initialize) {
+            initialize();
+        }
+
+        if (provider == null) {
+            throw new IllegalStateException("JSS has not yet been initialized");
+        }
+
+        return provider;
+    }
+
+    /**
+     * Fetches the JSS CryptoManager. If JSS has not yet been initialized, and the initialize
+     * argument is true, it will be initialized before fetching the provider. If the initialize
+     * argument is false and JSS is not initialized, this method throws an exception.
+     *
+     * @param initialize
+     *  whether or not to initialize JSS if it has not yet been initialized
+     *
+     * @throws JSSLoaderException
+     *  if the CryptoManager cannot be fetched
+     *
+     * @return
+     *  the JSS crypto manager
+     */
+    public static synchronized CryptoManager getCryptoManager(boolean initialize) {
+        if (initialize) {
+            initialize();
+        }
+
+        try {
+            // Impl note:
+            // This is necessary due to how we load JSS: The NotInitializedException moves packages
+            // between versions, and since we can't bind to a specific version, we can't hard-code
```

Comment:
I've added the git ref which makes the change, but it was changed in several versions simultaneously; so I'll let future us look at the tag list and make a decision based on that.

Reply:
That list that github shows for a commit, is all the tags that this commit exists in, from newest to oldest. so the version that it was introduced in first is the last one (v4.5.0, if you ignore all the a3/a4/b1 which I assume are alpha/beta tags)

Reply:
This coincides with the other reason we had to use reflection (the moving of the InitializationValues class). And 4.5.0+ is only used in RHEL8, so theoretically, if Satellite becomes RHEL8-only as is the current plan, we could remove both reflection workarounds

Reply:
That is unless and until JSS introduces another such breaking change (either between RHEL8 y-versions, or between RHEL8 and RHEL9), then we'd have to do something like this again, but it would obviously would be in different places in the code.

Reply:
Now that I think about it, another complication is what version is Hosted on. I believe they're on RHEL7, and we should get them on RHEL8 asap.

Reply:
I've changed the comment to only reference v4.5.